File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/connect_116.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.Node;

public class connect_116 {
    public Node connect(Node root) {
        if (root == null) return null;
        helper(root.left, root.right);
        return root;
    }

    private void helper(Node l, Node r) {
        if (l == null || r == null) return;
        l.next = r;
        helper(l.left, l.right);
        helper(r.left, r.right);
        helper(l.right, r.left);
    }
}

File: Clion/ZeroTrac/Tree/connect_116.cpp
//
// Created by TzashiNorpu on 10/25/2022.
//
#include "tree.h"

using namespace Tree;

Node *Solution::connect(Node *root) {
  function<void(Node *, Node *)> dfs = [&](Node *l, Node *r) -> void {
      if (!l) return;
      l->next = r;
      dfs(l->left, l->right);
      dfs(l->right, r->left);
      dfs(r->left, r->right);
  };
  if (root)
    dfs(root->left, root->right);
  return root;
}